File: dailyques/Q1458.py
#Q1458.py
class Solution(object):
    def maxDotProduct(self, nums1, nums2):
        n,m=len(nums1),len(nums2)
        NEG=-10**18
        dp=[[NEG]*(m+1) for _ in range(n+1)]
        for i in range(1,n+1):
            for j in range(1,m+1):
                prod=nums1[i-1]*nums2[j-1]
                dp[i][j]=max(prod, prod+dp[i-1][j-1], dp[i-1][j], dp[i][j-1])
        return dp[n][m]

File: dailyques/Q1458.cpp
//Q1458.cpp
class Solution {
public:
    int maxDotProduct(vector<int>& a, vector<int>& b) {
        int n=a.size(), m=b.size();
        const int NEG=-1000000000;
        vector<vector<int>> dp(n+1, vector<int>(m+1, NEG));
        for(int i=1;i<=n;i++){
            for(int j=1;j<=m;j++){
                int prod=a[i-1]*b[j-1];
                dp[i][j]=max({prod, prod+dp[i-1][j-1], dp[i-1][j], dp[i][j-1]});
            }
        }
        return dp[n][m];
    }
};
